### DISCUSSION TITLE
Why IceWM should support wayland

### DISCUSSION THREAD

**Discussion by CloisteredNeuron** [bbidulock/icewm#777] [General]
I'm a MS refugee whose WIN10 notebook will not be supported in about 16 months so I worked on creating a Debian XORG/IceWM installation.  I loved all 100MBs of it.  During the process of trying to configure my IceWM desktop to do what I had done in WIN10, very successfully, I discovered the X keystroke sharing feature.  This is the feature where you can run xinput (not as root) in test mode and capture keyboard intput into a calculator/notepad/sudo pw in another terminal window/financial site credentials in a browser.  Wayland is a solution but for MS refugees any wayland solution not supporting xwayland (GNOME & KDE support xwayland) will not run in VirtualBox very well so they can't get experience with lightweight wayland Linux desktops.  I've gotten labwc and sway to start on VirtualBox but they quickly lock up.  So for me, X is not an option.  From the MS world this keystroke sharing would be considered keylogging and those willing to switch from MS to Linux for security, would not view the X feature positively.

I’ve been stripping away KDE, specifically Debian plasma-desktop, components and have reduced RAM consumption to about 500MB but this is five times larger than my XORG/IceWM installation.  Maybe the maintainers of IceWM could reach out to KDE or GNOME to see if they could provide a stripped-down version of their products with the equivalent of what XORG on Debian provides.  IceWM could then port its window manager to a fleshed out wayland implementation that runs well in VirtualBox.  The upside for GNOME or KDE is they could say we already provide safer computing with our more full featured wayland-enabled desktop but with the introduction of IceWM on GNOME or IceWM on KDE they also offer the perfect lightweight environment.  Let’s say you want a VM to connect to financial sites and you want to isolate from your primary KDE/GNOME OS.  Why consume all the GNOME/KDE resources in a VM when all you need is to run a browser/notepad/calc/Libre Office – you just want to ensure that nifty AppImage calculator, created by who knows who, is not capturing your sudo pw and configuring a rat on your “secure” financial VM.

Another benefit of IceWM on KDE/GNOME infrastructure is the decision plaguing WIN10 users over which distro to choose is greatly simplified.  You hear again and again choose the Linux you’re comfortable with.  Simply sharing X security concerns take the 100s of distros to consider down to two if a WIN10 user wants to test a wayland Linux in VirtualBox.  IceWM comes in as a third option for less capable WIN10 systems.  My WIN10 system runs Debian XORG/IceWM VMs incredibly well, without browsing, in 300MB of total RAM.  I’m sure IceWM on KDE/GNOME infrastructure would require more resources than 300MB of total RAM but it will be much less than KDE/GNOME.  If the abandoned WIN10 users don’t continue to use unsupported WIN10 or switch to Linux we could potentially have millions of pounds of e-waste instead of functioning systems.  It seems INSANE to push away paying customers and provide no upgrade path.  Personally, I purchased two laptops in December 2016 in of all places a MS Store and less than five years later they say neither can be upgraded to WIN11.  These systems can run a WIN10 host, along with a WIN11 (lab mode configured) guest VM and a WIN10 guest VM concurrently, yet because the processor is not of the proper generation it is not capable of running WIN11.

**Comment by gijsbers** [comment]
Thanks a lot for your very informative writing. KDE/Gnome are big projects with many developers and many more resources, also financially. Compared to them IceWM is a tiny mouse sitting on an elephant.

I wonder if it is technically possible to disable the keyboard snooping. Could it be that disabling the XInputExtension would suffice? E.g. the following Xorg configuration:

```
Section "Extensions"
  Option "XInputExtension" "Disable"
EndSection
```

Maybe also disable the XTEST extension and the BIG-REQUESTS extension.

À propos, considering the many economic benefits you describe and the ambitious goals you assign to IceWM, could it be reasonable to help support the IceWM development?

**Comment by CloisteredNeuron** [comment]
I'll have to look into these extensions but from previous discussions sharing is just the way X was designed and was reasonable in the mid 80's.  In one conversation I put forth the notion that X functionality could be encapsulated into classes where a parent object would act as a traffic cop keeping track of each instantiated window and the currently active window.  The parent object would be the only object with access to keyboard input.  As keypress reached the parent it would send the keystrokes to the active window.  For those apps in an elevated access group (addgroup) which a user is willing to risk the inappropriate use of sensitive information, the parent would send all keystroke data to those apps.  The response to this was XORG would just not be interested as most people are interested in the new development and as you identified funding is an issue.  In another conversation about wayland I discovered every implementation of a windows manager using wayland has to reinvent the wheel.  That is, there is no wayland package to install like xorg.  It seems so much more reasonable to fix this glaring security issue with xorg and keep on using xorg.  But that is life.

I brought these ideas to the IceWM group as I was hoping GNOME or KDE might be more open to hear new ideas from a group which could directly help them.  Adding an IceWM flavor to their line-ups would elevate them to a go-to Linux as they have the features and security of their current wayland offerings and the simplicity of a light weight wayland windows manager which could run in VirtualBox.  For IceWM my hope is GNOME or KDE could provide a stripped-down version of their xwayland distros on which IceWM could be installed with minimal modification as compared to a port to wayland.  As you noted these organizations are much larger with more funding.  My hope is a GNOME/KDE partner could share some of this funding and human capital to make IceWM on GNOME/KDE a reality.  Maybe someday I'll have a non-profit to support such an effort but as of today I'm hoping to keep my 7.5-year-old notebook working for another 5 years.
